aptos-move/framework/aptos-framework/sources/permissioned_signer.move

```diff
@@ -0,0 +1,661 @@
+        let StorablePermissionedHandle::V1 {
+            master_account_addr,
+            permissions_storage_addr,
+            expiration_time: _
+        } = p;
+
+        assert!(
+            exists<GrantedPermissionHandles>(master_account_addr),
+            error::permission_denied(E_PERMISSION_REVOKED),
+        );
+        let granted_permissions =
+            borrow_global_mut<GrantedPermissionHandles>(master_account_addr);
+        let (found, idx) = vector::index_of(
+            &granted_permissions.active_handles, &permissions_storage_addr
+        );
+
+        // Removing the address from the active handle list if it's still active.
+        if(found) {
+            vector::swap_remove(&mut granted_permissions.active_handles, idx);
+        };
+
+        destroy_permissions_storage_address(permissions_storage_addr);
+    }
+
+    inline fun destroy_permissions_storage_address(
+        permissions_storage_addr: address
+    ) acquires PermissionStorage {
+        if (exists<PermissionStorage>(permissions_storage_addr)) {
+            let PermissionStorage::V1 { perms } =
+                move_from<PermissionStorage>(permissions_storage_addr);
+            simple_map::destroy(
+                perms,
+                |_dk| {},
+                |_dv| {}
+            );
+        }
+    }
+
+    /// Generate the permissioned signer based on the ephermeral permission handle.
+    ///
+    /// This signer can be used as a regular signer for other smart contracts. However when such
+    /// signer interacts with various framework functions, it would subject to permission checks
+    /// and would abort if check fails.
+    public fun signer_from_permissioned_handle(p: &PermissionedHandle): signer {
+        signer_from_permissioned_handle_impl(
+            p.master_account_addr, p.permissions_storage_addr
+        )
+    }
+
+    /// Generate the permissioned signer based on the storable permission handle.
+    public(package) fun signer_from_storable_permissioned_handle(
+        p: &StorablePermissionedHandle
+    ): signer {
+        assert!(
+            timestamp::now_seconds() < p.expiration_time,
+            error::permission_denied(E_PERMISSION_EXPIRED)
+        );
+        assert!(
+            exists<PermissionStorage>(p.permissions_storage_addr),
+            error::permission_denied(E_PERMISSION_REVOKED)
+        );
+        signer_from_permissioned_handle_impl(
+            p.master_account_addr, p.permissions_storage_addr
+        )
+    }
+
+    /// Revoke a specific storable permission handle immediately. This would disallow owner of
+    /// the storable permission handle to derive signer from it anymore.
+    public entry fun revoke_permission_storage_address(
+        s: &signer, permissions_storage_addr: address
+    ) acquires GrantedPermissionHandles, PermissionStorage {
+        assert!(
+            !is_permissioned_signer(s), error::permission_denied(ENOT_MASTER_SIGNER)
+        );
```

Comment:
assert_master_signer(s);

but actually better - should this be RevokePermissionHandlePermission ?